### ISSUE TITLE
Validations and TLS Endpoints Very Slow

### ISSUE THREAD

**Issue by crhuber** [kiali/kiali#4224]
**Describe the bug**
On a cluster with 1600 services and ~2000pods the `tls` and `validations` api endpoints take a long time to respond in large environment. For example tls takes 1.4 minutes to return. 
![image](https://user-images.githubusercontent.com/5857391/127179052-14ef8bfa-090a-4003-a0a7-ab0888b64cbd.png)

Valation endpoints result: {"errors":1,"objectCount":536,"warnings":73}. Load time is 30 seconds .

**Versions used**
Kiali UI
1.36.0 (ceb2b93f33f1c0445b51518691dcb48c2c801f8a)
Kiali Server
v1.36.0 (48c624aa9950b62b96c3c43591900b0291d483cf)
Kiali Container
v1.36.0
Components
Prometheus
2.19.2
Kubernetes
v1.20.4-eks-6b7464
Istio:
1.10.1

**To Reproduce**
Open Kiali> Open Overview Tab

**Expected behavior**

Overview, Applications, Workloads tab expected to load fast.

**Comment by jmazzitelli** [comment]
re: the validations, I created a PoC to disable the validations when configured so. The code is simple enough:

```diff
diff --git a/business/istio_validations.go b/business/istio_validations.go
index bc7d232d..8d7c1c29 100644
--- a/business/istio_validations.go
+++ b/business/istio_validations.go
@@ -217,13 +217,16 @@ func (in *IstioValidationsService) GetIstioObjectValidations(namespace string, o
 func runObjectCheckers(objectCheckers []ObjectChecker) models.IstioValidations {
        objectTypeValidations := models.IstioValidations{}
 
-       // Run checks for each IstioObject type
-       for _, objectChecker := range objectCheckers {
-               objectTypeValidations.MergeValidations(objectChecker.Check())
+       if config.Get().KialiFeatureFlags.Validations.Enable {
+               // Run checks for each IstioObject type
+               for _, objectChecker := range objectCheckers {
+                       objectTypeValidations.MergeValidations(objectChecker.Check())
+               }
+               objectTypeValidations.StripIgnoredChecks()
+       } else {
+               log.Warningf("Validations are disabled! You will not be told about any validation errors!")
        }
 
-       objectTypeValidations.StripIgnoredChecks()
-
        return objectTypeValidations
 }
 
diff --git a/config/config.go b/config/config.go
index 6663b1bd..7e07076d 100644
--- a/config/config.go
+++ b/config/config.go
@@ -351,6 +351,7 @@ type UIDefaults struct {
 // Validations defines default settings configured for the Validations subsystem
 type Validations struct {
        Ignore []string `yaml:"ignore,omitempty" json:"ignore,omitempty"`
+       Enable bool     `yaml:"enable,omitempty" json:"enable,omitempty"`
 }
```

But I'm not sure if doing a blanket disable is what we want. But, I point this out just to show the code that would be needed, and it is relatively easy.

We'd need a corresponding operator PR to add that setting to the Kiali CR, but that is equally easy to do:

```yaml
kiali_feature_flags:
  validations:
    ignore: []  # this is what we already have coming up in v1.38
    enable: true # we would add this, default is true, but you can set it to false
```

**Comment by xeviknal** [comment]
I do agree that before working on the https://github.com/kiali/kiali/discussions/4080 we could disable the validations for such a big clusters.

@jmazzitelli I think that what you propose can be done in a different way and avoiding kiali fetch unnecessarily resources and therefore wasting CPU and memory.

```diff
diff --git a/business/istio_validations.go b/business/istio_validations.go
index 8d4b9fbb..436a7b0f 100644
--- a/business/istio_validations.go
+++ b/business/istio_validations.go
@@ -25,6 +25,10 @@ type ObjectChecker interface {
 // GetValidations returns an IstioValidations object with all the checks found when running
 // all the enabled checkers. If service is "" then the whole namespace is validated.
 func (in *IstioValidationsService) GetValidations(namespace, service string) (models.IstioValidations, error) {
+       if !config.Get().KialiFeatureFlags.Validations.Enable {
+               return models.IstioValidations{}, nil
+       }
+
        // Check if user has access to the namespace (RBAC) in cache scenarios and/or
        // if namespace is accessible from Kiali (Deployment.AccessibleNamespaces)
        if _, err := in.businessLayer.Namespace.GetNamespace(namespace); err != nil {
@@ -123,6 +127,10 @@ func (in *IstioValidationsService) getAllObjectCheckers(namespace string, istioD
 }
 
 func (in *IstioValidationsService) GetIstioObjectValidations(namespace string, objectType string, object string) (models.IstioValidations, error) {
+       if !config.Get().KialiFeatureFlags.Validations.Enable {
+               return models.IstioValidations{}, nil
+       }
+
        var istioDetails kubernetes.IstioDetails
        var exportedResources kubernetes.ExportedResources
        var namespaces models.Namespaces
```

These two methods are the ones used at the business layer therefore are the ones called from the whole backend (or it should be).

**Comment by jmazzitelli** [comment]
> These two methods are the ones used at the business layer therefore are the ones called from the whole backend (or it should be).

OK. I was just looking for the common entry point (it looked like `runObjectCheckers` was it). As long as you catch all the different callers of `runObjectCheckers`, that's fine. We'll also want a warning message telling the user that all validations will be disabled (probably good to log that at boot time, now that I think about it - no need to keep spewing it).

**Comment by xeviknal** [comment]
Plus, I'd like to say that the real problem comes from the `tls` endpoint: which are not validations. Therefore, by disabling the validations this scenario will still have some performance problems.

The `tls` endpoints is in charge of checking which tls covering has each namespace of the system. And this one takes 1.4minutes. To calculate the tls status you need to consider:
- Mesh-wide mTlS: you need DRs and PeerAuthn which might live in all namespaces
- ns-wide mTLS: you need DRs and PeerAuthn which might live in all namespaces

Judging from the object count, this scenario has 536 istio objects in one single namespace. If you have more or less the same amount in each namespace and you have 10-100, there are lots of objects to check.

I don't know if instead of disabling the features we should just add a cache in front and make it expire every N minutes. TLS status shouldn't change so much.

**Comment by jmazzitelli** [comment]
Oh, then the validations isn't the problem then! We should look into making that tls endpoint more efficient.

**Comment by xeviknal** [comment]
I mean, validations are 30s, which is pretty big IMO.

We could do something like this to disable the tls endpoint.

```diff
diff --git a/business/tls.go b/business/tls.go
index 134786c7..b444c7ab 100644
--- a/business/tls.go
+++ b/business/tls.go
@@ -25,6 +25,10 @@ const (
 )
 
 func (in *TLSService) MeshWidemTLSStatus(namespaces []string) (models.MTLSStatus, error) {
+       if !config.Get().KialiFeatureFlags.TLS.Enable {
+               return models.MTLSStatus{}, nil
+       }
+
        pas, error := in.getMeshPeerAuthentications()
        if error != nil {
                return models.MTLSStatus{}, error
@@ -106,6 +110,10 @@ func (in *TLSService) getAllDestinationRules(namespaces []string) ([]kubernetes.
 }
 
 func (in TLSService) NamespaceWidemTLSStatus(namespace string) (models.MTLSStatus, error) {
+       if !config.Get().KialiFeatureFlags.TLS.Enable {
+               return models.MTLSStatus{}, nil
+       }
+
        pas, err := in.getPeerAuthentications(namespace)
        if err != nil {
                return models.MTLSStatus{}, nil

```

**Comment by jmazzitelli** [comment]
I agree with you @xeviknal that we should first look into making the performance better before we start disabling things like this. 

Let's therefore kick off a very concentrated effort to analyze the performance of this stuff and see if we can make them more efficient.

This is different from the case of wanting to ignore certain validations because that was OK in some cases (e.g. it may be the user has a valid use-case and they really aren't errors). In this case, it is just annoyingly slow. We should first try to make things faster.

**Comment by jshaughn** [comment]
The caching idea could be promising, as @israel-hdez mentioned in slack, it is like a pre-compute of information that rarely changes.

**Comment by xeviknal** [comment]
I can try to a cache for the mTLS endpoints. However, there will still be this 1.4min request everytime the cache expires. This could buy us some time. But I agree that before the cache we could quickly build an scenario to analyze the performance of the endpoint. If nothing found I'd go for cache or disabling. Disabling will for sure remove the trouble, give instant solution to the scenario.

**Comment by xeviknal** [comment]
@crhuber do you have automatic mTLS enabled? (it usually comes enabled by default)

**Comment by xeviknal** [comment]
Since the TLS endpoint deals with PeerAuthentications mainly (or should be), how many PeerAuthns/DRs do you have?

```bash
k get peerauthentications.security.istio.io --all-namespaces | wc -l
k get destinationrules.networking.istio.io --all-namespaces | wc -l
```

And regarding to the ns, how many do you have accessible to kiali?

I think this could give me a bit more of light for troubleshooting this.
Thanks

**Comment by crhuber** [comment]
here are the results

```
 k get peerauthentications.security.istio.io --all-namespaces | wc -l
No resources found
       0
 k get destinationrules.networking.istio.io --all-namespaces | wc -l
     458
k get namespace --show-labels  -l kiali-enabled=true | wc -l
      16
```

**Comment by lucasponce** [comment]
Side comment:

> Judging from the object count, this scenario has 536 istio objects in one single namespace. If you have more or less the same amount in each namespace and you have 10-100, there are lots of objects to check.

If there is not a hidden call to the backend, a figure like this shouldn't take that time unless there is something else involved.

But I agree with the comments, this issue is more related to the scability discussion https://github.com/kiali/kiali/discussions/4080 rather than a specific bug to address.

I don't share the approach to disable features as the first reaction to a scalability problem.

**Comment by xeviknal** [comment]
I did some research regarding to this slow down. The pain point here is that destination rules are the ones prevalent when setting up mTLS. Even if you are using autoMtls, DRs have the last word about how to establish the communication.
Therefore, Kiali needs to visit all the DRs in the cluster (scope of DRs are cluster-wide) in order to determine the mTLS status.
The magnitud of this deployment (458 DRs) makes this operarions pretty slow.

I have to experiment a little bit more with this but it looks like adding a first indexing of DRs by host name would speed up the search quite substantially. Accessing to the DRs that affects to either one namespace, the cluster or one service should take O(log n).

And if we cached the indexed DRs, those endpoints would definitely got faster.

This idea is not tested, not PoC. Just an idea.
